sdl-core/src/main/scala/io/smartdatalake/workflow/SubFeed.scala

```diff
@@ -92,109 +89,102 @@ trait SubFeedConverter[S <: SubFeed] {
+      val filterExpr = helper.col(partitionValuesColumn.head).isin(partitionValues.flatMap(_.elements.values):_*)
+      dataFrame.map(_.filter(filterExpr))
+    } else {
+      // filter with and/or expression if multiple partition columns
+      val filterExpr = PartitionValues.createFilterExpr(partitionValues)
+      dataFrame.map(_.filter(filterExpr))
     }
-    this.copy(dataFrame = dummyDf, isDummy = true)
-  }
-  override def applyExecutionModeResultForInput(result: ExecutionModeResult, mainInputId: DataObjectId)(implicit context: ActionPipelineContext): SparkSubFeed = {
-    // apply input filter
-    val inputFilter = if (this.dataObjectId == mainInputId) result.filter else None
-    this.copy(partitionValues = result.inputPartitionValues, filter = inputFilter, isSkipped = false).breakLineage // breaklineage keeps DataFrame schema without content
-  }
-  override def applyExecutionModeResultForOutput(result: ExecutionModeResult)(implicit context: ActionPipelineContext): SparkSubFeed = {
-    this.copy(partitionValues = result.inputPartitionValues, filter = result.filter, isSkipped = false, dataFrame = None)
-  }
+    // apply generic filter
+    val dfResult = if (filter.isDefined) dfPartitionFiltered.map(_.filter(helper.expr(filter.get)))
+    else dfPartitionFiltered
+    // return updated SubFeed
+    withDataFrame(dfResult)
+  }
+  def asDummy(): DataFrameSubFeed
+  def transform(transformer: GenericDataFrame => GenericDataFrame): DataFrameSubFeed = withDataFrame(dataFrame.map(transformer))
+  def movePartitionColumnsLast(partitions: Seq[String]): DataFrameSubFeed
 }
-object SparkSubFeed extends SubFeedConverter[SparkSubFeed] {
+trait DataFrameSubFeedCompanion extends SubFeedConverter[DataFrameSubFeed] {
+  protected def subFeedType: universe.Type
   /**
-   * This method is used to pass an output SubFeed as input SparkSubFeed to the next Action. SubFeed type might need conversion.
+   * This method can create the schema for reading DataObjects.
+   * If SubFeed subtypes have DataObjects with other methods to create a schema, they can override this method.
    */
-  override def fromSubFeed( subFeed: SubFeed )(implicit context: ActionPipelineContext): SparkSubFeed = {
-    subFeed match {
-      case sparkSubFeed: SparkSubFeed => sparkSubFeed.clearFilter() // make sure there is no filter, as filter can not be passed between actions.
-      case _ => SparkSubFeed(None, subFeed.dataObjectId, subFeed.partitionValues, subFeed.isDAGStart, subFeed.isSkipped)
+  def getDataObjectReadSchema(dataObject: DataObject with CanCreateDataFrame)(implicit context: ActionPipelineContext): Option[GenericSchema] = {
+    dataObject match {
+      case input: UserDefinedSchema if input.schema.isDefined =>
+        input.schema.map(dataObject.createReadSchema)
+      case input: SchemaValidation if input.schemaMin.isDefined =>
+        input.schemaMin.map(dataObject.createReadSchema)
+      case _ => None
     }
   }
+
+  /**
```

Comment:
This block can be put into another seperate trait SparkFunctions

Reply:
Done, created trait DataFrameFunctions